src/app/file-manager/file-explorer/file-explorer.component.ts

```diff
@@ -74,104 +64,146 @@ export class FileExplorerComponent implements OnInit {
-    this.countFiles();
+  clickOnRow(elem) {
+    if (elem.type === 'Folder') {
+      this.paginator.pageIndex = 0;
+      this.bucketObjects.data = [];
+      BucketService.delimiter = elem.delimiter;
+      const workspaceName = this.getWorkspaceName(elem.bucketName);
+      this.getBucketObjects(elem.bucketName, BucketService.delimiter, workspaceName);
+      this.breadcrumbPath = elem.prefix;
+    }
   }
 
-  toggleSelection() {
-    const newSelection = [];
-    this.files.forEach(x => {
-      if (this.selectedFiles.includes(x)) {
-        // do nothing
-      } else {
-        newSelection.push(x);
-      }
-    });
-    this.selectedFiles = newSelection;
-    this.countFiles();
+  pathChanged(event) {
+    this.paginator.pageIndex = 0;
+    let delimiter = event.label !== event.path ? this.getDelimiter(event.label) : this.DELIMITER;
+    const workspace = this.breadcrumbPath.substr(0, this.breadcrumbPath.indexOf(this.DELIMITER));
+    const bucket = FirecloudApiService.workspaces.get(workspace);
+    this.bucketObjects.data = [];
+    BucketService.delimiter = delimiter;
+    this.getBucketObjects(bucket, delimiter, workspace);
+    delimiter = delimiter.substring(delimiter.length - 1, delimiter.length) === this.DELIMITER
+                ? delimiter.substr(0, delimiter.lastIndexOf(this.DELIMITER)) : delimiter;
+    this.breadcrumbPath = workspace + this.DELIMITER + delimiter;
   }
 
-  private selectRecursive(node: TreeNode, isExpand: boolean) {
-    this.selectedFiles = [...this.selectedFiles, node];
-    if (node.children) {
-      node.children.forEach(childNode => {
-        this.selectRecursive(childNode, isExpand);
+  getBucketObjects(bucket: String, delimiter: String, workspace: String) {
+    this.data = this.bucketService.getBucketData(bucket, delimiter, null, workspace).subscribe(
+      elements => {
+        this.bucketObjects.data = elements;
       });
-    }
   }
 
   selectionDone() {
-    const items = {selectedFiles: this.selectedFiles, totalSize: this.totalSize };
+    const items = { selectedFiles: this.selection.selected, totalSize: this.totalSize };
     const dialogRef = this.dialog.open(FileDownloadModalComponent, {
       width: '500px',
       disableClose: true,
       data: items
-
     });
   }
 
   disableButton() {
-    return this.downloadInProgress || this.fileCount <= 0;
+    return this.downloadInProgress || this.selection.selected.length === 0;
+  }
+
+  isSelected(event, row) {
+    const selectedItem = this.selection.selected.find(item => item.id === row.id);
+    if (event.checked && selectedItem === undefined) {
+      this.selection.toggle(row);
+    } else if (selectedItem !== undefined) {
+      this.selection.deselect(selectedItem);
+    }
+    this.countFiles();
+  }
+
+  getWorkspaceName(bucketName) {
+    let workspaceName = '';
+    FirecloudApiService.workspaces.forEach((value: string, key: string) => {
+      if (value === bucketName) {
+        workspaceName = key;
```

Comment:
Once you find the matching workspace, you can `break;` to avoid checking more workspaces.